File: utils/read_binary_data/matlab/read_restart_file.m
% -
%
% SPDX-FileCopyrightText: Pedro Costa and the CaNS contributors
% SPDX-License-Identifier: MIT
%
% -
%
% setting up some parameters
%
precision = 'double';     % precision of the real-valued data
r0 = [0.,0.,0.];    % domain origin
%
% read geometry file
%
geofile  = "geometry.out";
data = dlmread(geofile);
ng   = data(1,:);
l    = data(2,:);
dl   = l./ng;
%
% read and generate grid
%
xp = linspace(r0(1)+dl(1)/2.,r0(1)+l(1)-dl(1)/2.,ng(1)); % centered  x grid
yp = linspace(r0(2)+dl(2)/2.,r0(2)+l(2)-dl(2)/2.,ng(2)); % centered  y grid
zp = linspace(r0(3)+dl(3)/2.,r0(3)+l(3)-dl(3)/2.,ng(3)); % centered  z grid
xu = xp + dl(1)/2.; % staggered x grid
yv = yp + dl(2)/2.; % staggered y grid
zw = zp + dl(3)/2.; % staggered z grid
if(exist('grid.bin','file'))
    f   = fopen('grid.bin');
    grid_z = fread(f,[ng(3),4],precision);
    fclose(f);
    zp = r0(3) + grid_z(:,3)'; % centered  z grid
    zw = r0(3) + grid_z(:,4)'; % staggered z grid
end
%
% read checkpoint binary file
%
filenamei = input("Name of the binary restart file written by CaNS [fld.bin]: ");
if isempty(filenamei)
    filenamei = "fld.bin";
end
data       = zeros([ng(1),ng(2),ng(3),4]); % u(:,:,:),v(:,:,:),w(:,:,:),p(:,:,:)
fldinfo    = zeros([2,1]);
[filepath,name,ext] = fileparts(filenamei);
fieldnames = {'u','v','w','p'};
prefix = fullfile(filepath,name);
for q = 1:4
    suffix = ['_',fieldnames{q}];
    if endsWith(name,suffix)
        prefix = fullfile(filepath,name(1:end-length(suffix)));
    end
end
split_files = cell(1,4);
is_split = true;
for q = 1:4
    split_files{q} = [prefix,'_',fieldnames{q},'.bin'];
    is_split = is_split && exist(split_files{q},'file');
end
if is_split
    for q = 1:4
        f = fopen(split_files{q});
        data(:,:,:,q) = reshape(fread(f,ng(1)*ng(2)*ng(3),precision),[ng(1),ng(2),ng(3)]);
        if q == 1
            fldinfo(:) = fread(f,2,precision);
        end
        fclose(f);
    end
else
    f = fopen(filenamei);
    for q = 1:4
        data(:,:,:,q) = reshape(fread(f,ng(1)*ng(2)*ng(3),precision),[ng(1),ng(2),ng(3)]);
    end
    fldinfo(:) = fread(f,2,precision);
    fclose(f);
end
%
% store data in arrays
%
u = data(:,:,:,1);
v = data(:,:,:,2);
w = data(:,:,:,3);
p = data(:,:,:,4);
time  =       fldinfo(1);
istep = round(fldinfo(2));

File: utils/read_binary_data/python/read_restart_file.py
# -
#
# SPDX-FileCopyrightText: Pedro Costa and the CaNS contributors
# SPDX-License-Identifier: MIT
#
# -
#!/usr/bin/env python
def read_restart_file(filenamei):
    import os
    import numpy as np
    #
    # setting up some parameters
    #
    iprecision = 8            # precision of the real-valued data
    r0 = np.array([0.,0.,0.]) # domain origin
    precision  = 'float64'
    if(iprecision == 4): precision = 'float32'
    #
    # read geometry file
    #
    geofile  = "geometry.out"
    data = np.loadtxt(geofile, comments = "!", max_rows = 2)
    ng = data[0,:].astype('int')
    l  = data[1,:]
    dl = l/(1.*ng)
    #
    # read and generate grid
    #
    xp = np.linspace(r0[0]+dl[0]/2.,r0[0]+l[0]-dl[0]/2.,ng[0]) # centered grid
    yp = np.linspace(r0[1]+dl[1]/2.,r0[1]+l[1]-dl[1]/2.,ng[1]) # centered grid
    zp = np.linspace(r0[2]+dl[2]/2.,r0[2]+l[2]-dl[2]/2.,ng[2]) # centered grid
    xu = xp + dl[0]/2. # staggered grid
    yv = yp + dl[1]/2. # staggered grid
    zw = zp + dl[2]/2. # staggered grid
    if(os.path.exists('grid.bin')):
        f   = open('grid.bin','rb')
        grid_z = np.fromfile(f,dtype=precision)
        f.close()
        grid_z = np.reshape(grid_z,(ng[2],4),order='F')
        zp = r0[2] + grid_z[:,2] # centered  z grid
        zw = r0[2] + grid_z[:,3] # staggered z grid
    #
    # read checkpoint binary file
    #
    disp = np.prod(ng)
    fldinfo = np.zeros([2])
    legacy_size = iprecision*(4*disp+2)
    def split_prefix(filename):
        root, ext = os.path.splitext(filename)
        if(ext == ".bin"):
            for suffix in ["_u","_v","_w","_p"]:
                if(root.endswith(suffix)):
                    return root[:-len(suffix)]
            return root
        return filename
    def read_one(filename):
        with open(filename,'rb') as f:
            fld = np.fromfile(f,dtype=precision,count=disp)
            info = np.fromfile(f,dtype=precision,count=2)
        return np.reshape(fld,(ng[0],ng[1],ng[2]),order='F'), info
    if(os.path.exists(filenamei) and os.path.getsize(filenamei) == legacy_size):
        #
        # u, v, w, p
        #
        data = np.zeros([ng[0],ng[1],ng[2],4])
        offset = 0
        with open(filenamei,'rb') as f:
            for q in range(4):
                f.seek(offset)
                fld = np.fromfile(f,dtype=precision,count=disp)
                data[:,:,:,q] = np.reshape(fld,(ng[0],ng[1],ng[2]),order='F')
                offset += iprecision*disp
            f.seek(offset)
            fldinfo[:] = np.fromfile(f,dtype=precision,count=2)
        u = data[:,:,:,0]
        v = data[:,:,:,1]
        w = data[:,:,:,2]
        p = data[:,:,:,3]
    else:
        prefix = split_prefix(filenamei)
        u, fldinfo = read_one(prefix+"_u.bin")
        v, _       = read_one(prefix+"_v.bin")
        w, _       = read_one(prefix+"_w.bin")
        p, _       = read_one(prefix+"_p.bin")
    time  =     fldinfo[0]
    istep = int(fldinfo[1])
    return u, v, w, p, time, istep

if __name__ == "__main__":
    filenamei = input("Name of the binary restart file written by CaNS [fld.bin]: ") or "fld.bin"
    u, v, w, p, time, istep = read_restart_file(filenamei)
